nntrainer/models/neuralnet.cpp

```diff
@@ -577,20 +576,19 @@ int NeuralNetwork::train_run() {
     iter = 0;
   }
 
-  auto &first_layer = model_graph.getSortedLayerNode(0)->getObject();
-  auto &last_layer =
-    model_graph.getSortedLayerNode(model_graph.size() - 1)->getObject();
+  auto const &first_layer_node = model_graph.getSortedLayerNode(0);
+  auto const &last_layer_node =
+    model_graph.getSortedLayerNode(model_graph.size() - 1);
 
-  auto &output = last_layer->net_hidden[0]->getVariableRef();
-  auto &label = last_layer->net_hidden[0]->getGradientRef();
-  auto &in = first_layer->net_input[0]->getVariableRef();
+  auto &output = last_layer_node->getOutput(0);
+  auto &label = last_layer_node->getOutputGrad(0);
+  auto &in = first_layer_node->getInput(0);
 
   /// @todo migrate this to trait based system; sth like need label?
 
   for (auto iter = model_graph.cbegin(); iter != model_graph.cend(); iter++) {
-    auto const &layer_ = (*iter)->getObject();
-    if (layer_->getType() == "centroid_knn") {
-      layer_->net_hidden[0]->getGradientRef() = label;
+    if ((*iter)->getType() == "centroid_knn") {
```

Comment:
```suggestion
    if ((*iter)->getObject()->requireLabel()) {
```

Reply:
consider for a later PR.

Reply:
This is merged, @lhs8928  could you cherry pick https://github.com/nnstreamer/nntrainer/pull/1263 to make the PR complete?

Reply:
Done